crates/std_detect/src/detect/cache.rs

```diff
@@ -100,40 +98,52 @@ impl Cache {
     const fn uninitialized() -> Self {
         Cache(AtomicUsize::new(usize::MAX))
     }
-    /// Is the cache uninitialized?
-    #[inline]
-    pub(crate) fn is_uninitialized(&self) -> bool {
-        self.0.load(Ordering::SeqCst) == usize::MAX
-    }
 
-    /// Is the `bit` in the cache set?
+    /// Is the `bit` in the cache set? Returns `None` if the cache has not been initialized.
     #[inline]
-    pub(crate) fn test(&self, bit: u32) -> bool {
-        test_bit(self.0.load(Ordering::SeqCst) as u64, bit)
+    pub(crate) fn test(&self, bit: u32) -> Option<bool> {
+        let cached = self.0.load(Ordering::Relaxed);
+        if cached == usize::MAX {
+            None
+        } else {
+            Some(test_bit(cached as u64, bit))
+        }
     }
 
     /// Initializes the cache.
     #[inline]
-    fn initialize(&self, value: usize) {
+    fn initialize(&self, value: usize) -> usize {
+        // Use `SeqCst` store to ensure a thread which sees one entry in `CACHE`
+        // initialized will definitely see the other. While it likely never can
+        // happen, this avoids something like:
+        //
+        // ```
+        // is_foo_feature_detected!("bar") && is_foo_feature_detected!("baz")
+        // ```
+        //
+        // needing to initialize the cache twice.
         self.0.store(value, Ordering::SeqCst);
```

Comment:
This can be `Relaxed`. `SeqCst` here only ensures an ordering if the loads are also `SeqCst`.

Reply:
Hmm, my thought at the time that it would guarantee consistency with the other store, and that it would tend to force a fence that would flush the writes, which would help ensure (but not guarantee) that the relaxed loads would see the stores sooner.

But today thinking about it the first bit of this doesn't really make sense.

That said, the comment that we can use relaxed because "we are only interested in the effects of operations on a single memory location" is:

1. Sort of just wrong because there are two cache cells, admittedly they're somewhat independent.
2. Misleading, because Relaxed behaving atomically (unless I'm mistaken?) doesn't mean relaxed stores are visible in relaxed loads on all threads.

Neither of these matter for correctness — initializing the cache is an idempotent operation...

But if we're doing non-x86 initialization that has to read e.g. `/proc/cpuinfo`, we'd *really* like to not initialize twice (Maybe even enough that that code should take a lock?)

I think in a follow-up I'd like to change `detect_and_initialize` to do a double-check using an `Acquire` read, and finally initialize with a `Release` store. This would look like: https://gist.github.com/thomcc/168348bfab306439bdc1833750d2b929 and I can add it to this PR if you want.

Reply:
I don't think it's such a big deal if we end up reading `/proc/cpuinfo` twice. It will rarely happen in practice and it's not *that* slow that we can't afford to do it once per thread.

Double-checking with an `Acquire` is useless. Acquire and release are memory *ordering* constraints, they have nothing to do with cache coherency: all they do is ensure that operations on one CPU are performed in a particular order.

Reply:
> they have nothing to do with cache coherency: all they do is ensure that operations on one CPU are performed in a particular order.

Hmm... Really? My understanding here has always been that when you do, say, a `a.store(x, Release)`, this needs to:

- Tell the optimizer not to reorder the writes.
- Tell the CPU that the writes must be done in that order.
- Tell the cache contoller on the chip that the cache lines impacted by the memory locations in question must be flushed in that order.

With steps 2 and 3 done via either fencing or instruction selection, of course.

Certainly on older hardware (now-irrelevant) it was possible to read a pointer, and then read a value through that pointer, and have the value be older than the pointer (and IIUC this is why C++'s `memory_order_consume` exists and is stronger than `memory_order_relaxed` — if the first load is performed with consume it's sufficient, but not with relaxed), and this would happen because the cache lines wouldn't be flushed in order (which is the case on now-irrelevant machines like the Alpha).

IIUC this is also why false sharing and cache line padding is important for performance in practice — most existing CPUs implement atomic operations via the same protocols that they use for maintaining cache-coherency.

That's all/mostly hardware specific... but, the C++ standard mentions:

> Implementations should make atomic stores visible to atomic loads within a reasonable amount of time.

Which definitely implies that there's no guarantee that relaxed reads get a "current" value.

Also, for `Acquire`, the cppreference page cited by the `Ordering` docs says:

> All writes in other threads that release the same atomic variable are visible in the current thread

Wich is why I considered switching to Acquire/Release and did a double-check there acquire — to guarantee that the most recent write would be visible in this thread.

---

That said I don't care that much, I'm not married to doing a double check there, but I don't think it's right to say it's useless, or that ordering has nothing to do with cache coherency.

But maybe I'm still confused?

Reply:
>> Implementations should make atomic stores visible to atomic loads within a reasonable amount of time.
>
> Which definitely implies that there's no guarantee that relaxed reads get a "current" value.

This applies to *all* atomic stores, not just relaxed ones. Basically it means CPUs shouldn't keep a store in a store buffer indefinitely.

----

You could implement your double-check with only `Relaxed` atomics and it would have the same effect. An acquire operation basically means "any loads or stores after the acquire operation cannot be reordered before the acquire operation". I suggest reading [this](https://preshing.com/20120913/acquire-and-release-semantics/) article which has a good explanation of acquire and release semantics.

Reply:
> An acquire operation basically means "any loads or stores after the acquire operation cannot be reordered before the acquire operation".

Isn't an acquire read guaranteed to *also* guaranteed get the value from a previous release store?

That blog (which I've read before) this has a really similar example, and says

> For now, suffice to say that for this technique to work in general, the acquire and release semantics must apply to the same variable – in this case, Ready – and both the load and store must be atomic operations.

It also links https://preshing.com/20130823/the-synchronizes-with-relation/ which further says that the Acquire load will see the Release store, because of the "synchronizes with" guarantee, which says the load will take "its value from any side effect in the release sequence headed by A.", which I had seen before but had to dig up:

> A release sequence headed by a release operation A on an atomic object M is a maximal contiguous sub-sequence of side effects in the modification order of M, where the first operation is A, and every subsequent operation
>
> - is performed by the same thread that performed A
> - is an atomic read-modify-write operation.

In our case, the release sequence is just the release store in initialize, since it's the only side effect that's ever performed on the atomic. So it seems like the `Acquire` load would be guaranteed to read the value written by a `Release` store.

Unless you're saying that `Relaxed` is sufficient to guarantee this too (which... maybe? I have a *lot* of trouble confirming this for anything, except for maybe where it's part of an RMW operation, which it's not here).

Sorry for dragging this on, BTW, I'm not trying to bicker or anything — I'm just unsure why `Acquire`/`Release` would be equivalently insufficient to `Relaxed`, or why `Relaxed` *is* sufficient if you're saying it is.

Reply:
> Isn't an acquire read guaranteed to also guaranteed get the value from a previous release store?

An acquire read doesn't make any guarantees about itself, it makes guarantees about *other* loads and stores relative to the acquire operation. Specifically that *if* an acquire read returns a value X that was release-stored then any loads/stores after the acquire are guaranteed to "happen before" any loads stores that were before the release.

The classic example here is a spinlock: an acquire operation to grab the lock ensures that any loads/stores afterwards are not reordered to happen before the lock is acquired. Similarly a release store to release the lock ensures that any prior loads/stores are not reordered to happen after the lock is released.

> Unless you're saying that Relaxed is sufficient to guarantee this too (which... maybe? I have a lot of trouble confirming this for anything, except for maybe where it's part of an RMW operation, which it's not here).

Cache coherency protocols ensure that any load instruction will always try to read the latest version of a value. It does not matter whether that load is `Acquire` or `Relaxed`. Memory orderings are only there to restrict the order of *other* load/stores relative to an atomic operation. Practically speaking this is implemented in CPUs by preventing some out-of-order executions that would result in the memory seeing the operations in the "wrong" order.